### ISSUE TITLE
[HELP] De-parallelize reward function

### ISSUE THREAD

**Issue by stanleyjs** [Acellera/acegen-open#53]
Hello,

I am attempting to use a reward function that is an instance of another model. I've got this implemented in MolScore. However, when I try to run reinvent with this, it attempts to initialize the reward model in parallel as many jobs. Is there a way to use a single instance of a scoring function?

thank you

**Comment by MorganCThomas** [comment]
I need a lot more detail, is there any way I can reproduce this? Also can I double checking you're referring to the `reinvent.py` script in acegen-open.

I'm not even clear if this is an issue related to acegen-open (just the reinvent.py script) or MolScore?

**Comment by stanleyjs** [comment]
Yes, I'm trying to use reinvent. haven't verified on other RL algorithms.  I'm also not clear if it's MolScore or reinvent, though it appears to be Acegen as I'm seeing a lot of the logging from repetitive molscore imports (as if they are being imported in subprocesses).

Because of the way molscore is engineered in with Acegen, reproducing this takes a lot of overhead, but I can give you what I'm using. This could probably be pared down into a smaller MWE by replacing the admet_ai.py with a hello world scoring.  You'll needd to remove the .txt file extensions as GitHub freaked out when I tried to upload .py and .yaml.

this is the yaml for the reinvent run
[config_denovo.yaml.txt](https://github.com/user-attachments/files/17804165/config_denovo.yaml.txt)

These are the required files for molscore. note that they run fine for scoring without acegen involved.
[__init__.py.txt](https://github.com/user-attachments/files/17804179/__init__.py.txt)
[admet_ai.py.txt](https://github.com/user-attachments/files/17804166/admet_ai.py.txt)
this is the config file for the scoring
[toxicity.json](https://github.com/user-attachments/files/17804145/toxicity.json)

**Comment by stanleyjs** [comment]
bug appears using reinforce as well

**Comment by MorganCThomas** [comment]
One thing to note is that torchrl sets multiprocessing to 'spawn' instead of default 'fork' and this doesn't play nicely with MolScore. Hence why in MolScore whenever I use multiprocessing I explicitly specify the context with `mp = multiprocessing.get_context('fork')`. If this isn't done I see the echoing behaviour you've described. 

I suspect this is because admet_ai is internally using multiprocessing which is where it goes crazy. As for the MWE, that shouldn't be echoing.

On another note, I've implemented ADMETAI myself already as a seperate server to avoid issues like this (which I generally do running ML models as scoring functions). I realised I never put it in `__init__.py` so I've just pushed that to the develop branch. Have you tried that?

**Comment by stanleyjs** [comment]
@MorganCThomas Thanks - yeah admet is using multiprocessing in two places - fingerprint calculation and in the torch dataloader for feeding the model. 

I noticed you have an implementation of ADMETAI as a server but I was having issues getting it to respond to queries in a timely fashion after the first query (just inside of molscore using .score, no other pipeline components). It was getting hung so I took the easy path around and implemented it as a class. Are you able to get it to work with your implementation?
